### ISSUE TITLE
`nextstrain view` error message with standalone CLI

### ISSUE THREAD

**Issue by joverlee521** [nextstrain/cli#242]
Using **standalone** CLI version 6.0.0 on macOS 10.14.6, running `nextstrain view auspice` outputs this error message:
```
Traceback (most recent call last):
  File "multiprocessing.spawn", line 115, in spawn_main
TypeError: 'dict' object cannot be interpreted as an integer
usage: nextstrain [-h] {build,view,deploy,remote,shell,update,setup,check-setup,login,logout,whoami,version,init-shell,authorization,debugger} ...
nextstrain: error: argument {build,view,deploy,remote,shell,update,setup,check-setup,login,logout,whoami,version,init-shell,authorization,debugger}: invalid choice: 'from multiprocessing.resource_tracker import main;main(3)' (choose from 'build', 'view', 'deploy', 'remote', 'shell', 'update', 'setup', 'check-setup', 'login', 'logout', 'whoami', 'version', 'init-shell', 'authorization', 'debugger')
```
After the error output, Auspice still runs as expected but the web browser does not open automatically.

**Comment by tsibley** [comment]
Oh bother. I see what's happening here, and I can reproduce it on Linux.

The issue is that `multiprocessing`'s internals seem to expect `sys.executable` to be the standard Python interpreter, with all its standard command-line argument behaviour. In the case of our standalone version, `sys.executable` is _not_ the standard interpreter, but `nextstrain` itself.

I'm sure there's something we can do here.

**Comment by tsibley** [comment]
(I didn't notice this when developing the new `view` functionality because I wasn't testing locally with the standalone binary.)

**Comment by tsibley** [comment]
I saw this warning in the multiprocessing docs, but didn't register that it would be an issue also for PyOxizider-produced binaries.

![Screenshot from 2022-12-16 10-15-03](https://user-images.githubusercontent.com/79913/208162902-4d9cab3c-ebf7-44ca-8582-768f67b899dc.png)

**Comment by tsibley** [comment]
We can [change the executable to use with `multiprocessing.set_executable()`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_executable), but that's not sufficient on its own as we don't have a way to invoke a standalone Python interpreter when running under the standalone binary.

I think we could make multiprocessing work here, but I'm wary of if it's worth it and might step back and choose another solution instead.

**Comment by tsibley** [comment]
For context, the whole reason we're spawning another process before running `webbrowser.open()` is that the latter might block and wait on some platforms or on some browsers. We don't want that, as we always want to launch the browser in the background and then exec into Auspice. (For structural reasons in the existing codebase, I don't want to flip the order of those around.)

**Comment by tsibley** [comment]
> the whole reason we're spawning another process before running `webbrowser.open()` is that the latter might block and wait on some platforms or on some browsers. We don't want that, as we always want to launch the browser in the background

I dug into this assumption of mine that `webbrowser.open()` potentially blocking on some browsers to exit could stall our main thread. That assumption lead me to use separate processes (no shared GIL) instead of separate threads (shared GIL).

However, the assumption was wrong! While some of `webbrowser`'s various classes do call `subprocess.Popen.wait()` and that blocks in Python-land, the underlying `os.waitpid()` call being made by `subprocess` (at least on Unix systems) releases the Python GIL when it calls the libc `waitpid()`.

https://github.com/python/cpython/blob/a757c1b7d3dfec875be15f8bd83777142d75b376/Modules/posixmodule.c#L8429-L8431

This means other Python threads can continue. In retrospect, it would be a weird oversight if the GIL wasn't released in this case, though it's conceivable technical constraints could require it I guess.

I also empirically verified my new understanding:

```python
from threading import Thread
from webbrowser import GenericBrowser
from time import time, sleep

start = time()

def browser():
    print("browser starting")
    GenericBrowser("sleep").open("2")
    print("browser done in", time() - start)

Thread(target=browser).start()

while time() - start < 4:
    print("main thread sleeping for 1s")
    sleep(1)
print("main thread done in", time() - start)
```

```
$ time python3 /tmp/wait
browser starting
main thread sleeping for 1s
main thread sleeping for 1s
browser done in 2.0021848678588867
main thread sleeping for 1s
main thread sleeping for 1s
main thread done in 4.00590705871582

real    0m4.103s
user    0m0.069s
sys     0m0.033s
```

If all threads were blocked, the total time would have been ~6s (4 + 2) instead of ~4s (max(4, 2)) and the output wouldn't have been interleaved.

Still need to verify the same behaviour on Windows too. Can only examine CPython at the moment, not run code on Windows yet, but will see about that later.

**Comment by tsibley** [comment]
Windows appears to have the same behaviour, though a different call chain.

Unix (really POSIX) systems calling `subprocess.Popen.wait(None)` go thru [`._wait(None)`](https://github.com/python/cpython/blob/a757c1b7d3dfec875be15f8bd83777142d75b376/Lib/subprocess.py#L1911-L1949) then `._try_wait(0)` then `os.waitpid(..., 0)`. 0 means "no `waitpid()` flags", i.e. no `os.WNOHANG` as used when timeout is not `None`, i.e. do a blocking operation. `os.waitpid()` releases the GIL when calling `waitpid()` on POSIX systems (or Windows systems with `_cwait()`, but Windows systems calling `Popen.wait()` don't go thru `os.waitpid()`, see below).

Windows systems calling `subprocess.Popen.wait(None)` go thru [a _different_ `._wait(None)`](https://github.com/python/cpython/blob/a757c1b7d3dfec875be15f8bd83777142d75b376/Lib/subprocess.py#L1482-L1495) and then `_winapi.WaitForSingleObject(..., _winapi.INFINITE)` which does a blocking operation, but also [releases the GIL while doing so.](https://github.com/python/cpython/blob/a757c1b7d3dfec875be15f8bd83777142d75b376/Modules/_winapi.c#L1881-L1883)

**Comment by tsibley** [comment]
Ok, so the remaining issue that I'm not sure we can start a new thread and then replace the main thread with another program using `exec()` without the browser thread going poof or otherwise misbehaving.

**Comment by tsibley** [comment]
Yeah, we can't. One workaround to keep using `multiprocessing`, I think, would be to use the default start method (different per platform) instead of always using `spawn`:

https://github.com/nextstrain/cli/blob/f19e50198331d1d51492b5a75f9416d7c7441591/nextstrain/cli/command/view.py#L63-L71

That might make that code a bit more brittle, but it should be doable.

Also, it turns out that [PyOxidizer has special-handling for `multiprocessing`](https://pyoxidizer.readthedocs.io/en/stable/pyoxidizer_packaging_multiprocessing.html) that we're not getting when we don't use the default start method.

I'll digest those docs and look into what's best here for us.

**Comment by tsibley** [comment]
PyOxidizer has support for the `spawn` start method (i.e. it can detect when it's being launched for a new `multiprocessing` process)... but only on Windows. orz

So I think we have to use `fork`, which means we just need to be careful not to break `spawn` on Windows by being careful about scopes and the variables that will be available to `spawn`-started processes vs. `fork`-started processes.

**Comment by tsibley** [comment]
Ah, seems like the issue with `spawn` and PyOxidizer is this bug: https://github.com/indygreg/PyOxidizer/issues/603

and I think we could work around it by explicitly calling `multiprocessing.get_context("spawn").freeze_support()` in our `main()`? Will test.

It seems the [`multiprocessing` docs on `freeze_support()`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.freeze_support) are a bit misleading, as they say:

> Calling `freeze_support()` has no effect when invoked on any operating system other than Windows. In addition, if the module is being run normally by the Python interpreter on Windows (the program has not been frozen), then `freeze_support()` has no effect.

but that's [only true for the `fork` and `forkserver` start methods](https://github.com/python/cpython/blob/2dda6680ccb50bd1d8a8dcf0d524feb7daf5a3af/Lib/multiprocessing/context.py#L144-L150)! The [`spawn` start method has a `freeze_support()` that DTRT regardless of platform](https://github.com/python/cpython/blob/2dda6680ccb50bd1d8a8dcf0d524feb7daf5a3af/Lib/multiprocessing/spawn.py#L62-L75).

**Comment by tsibley** [comment]
Ahh, no, the multiprocessing docs are correct! I misread the code. `multiprocessing.spawn.freeze_support()` is a module function not a class method (e.g. on the `SpawnContext` class returned by `multiprocessing.get_context("spawn")`).

**Comment by tsibley** [comment]
Ok, so this patch:

```diff
diff --git a/nextstrain/cli/__main__.py b/nextstrain/cli/__main__.py
index 3dbda8b..8af3c85 100644
--- a/nextstrain/cli/__main__.py
+++ b/nextstrain/cli/__main__.py
@@ -2,6 +2,7 @@
 Stub function and module used as a setuptools entry point.
 """
 
+import multiprocessing.spawn
 import sys
 from io import TextIOWrapper
 from sys import argv, exit
@@ -16,6 +17,17 @@ def main():
     sys.stdout = reconfigure_stdio(sys.stdout) # type: ignore[arg-type]
     sys.stderr = reconfigure_stdio(sys.stderr) # type: ignore[arg-type]
 
+    # If we're frozen (e.g. by PyOxidizer into our standalone binary), then
+    # enable support for multiprocessing's "spawn" start method, which we use
+    # in nextstrain/cli/command/view.py.
+    #
+    # PyOxidizer would normally do this for us automatically, but its
+    # re-implementation is buggy¹, so we disble it in pyoxidizer.bzl.
+    #
+    # ¹ https://github.com/indygreg/PyOxidizer/issues/603
+    if getattr(sys, "frozen", None):
+        multiprocessing.spawn.freeze_support()
+
     return cli.run( argv[1:] )
 
 
diff --git a/pyoxidizer.bzl b/pyoxidizer.bzl
index f0ced4a..9d93041 100644
--- a/pyoxidizer.bzl
+++ b/pyoxidizer.bzl
@@ -64,6 +64,19 @@ def make_exe():
     # we're in a standalone installation or not at runtime.
     python_config.x_options = ["nextstrain-cli-is-standalone"]
 
+    # Work around a PyOxidizer bug¹ in its multiprocessing "spawn" support by
+    # disabling its reimplementation of multiprocessing.freeze_support().
+    # We'll call it ourselves.
+    #
+    # ¹ https://github.com/indygreg/PyOxidizer/issues/603
+    python_config.multiprocessing_auto_dispatch = False
+
+    # Don't automatically set a start method.  We'll do it ourselves as
+    # appropriate.  The automatically set start method is most useful when
+    # multiprocessing_auto_dispatch also works (so that the start method can be
+    # completely transparent to the app) but alas it doesn't (and isn't).
+    python_config.multiprocessing_start_method = "none"
+
     # Produce a PythonExecutable from a Python distribution, embedded
     # resources, and other options. The returned object represents the
     # standalone executable that will be built.
```

Gets past the original issue of spawning new multiprocessing processes with our standalone executable...

```
TypeError: 'dict' object cannot be interpreted as an integer
```

but runs squarely into issues spawning the "resource tracker" process needed by multiprocessing on POSIX systems:

```
invalid choice: 'from multiprocessing.resource_tracker import main;main(4)'
```

(which was the next error in the original report above).

The resource tracker is run _without_ the `--multiprocessing-fork` flag detected by `freeze_support()`:

https://github.com/python/cpython/blob/2dda6680ccb50bd1d8a8dcf0d524feb7daf5a3af/Lib/multiprocessing/resource_tracker.py#L112-L119

or an equivalent.

**Comment by tsibley** [comment]
This little proof-of-concept hack demonstrates that it should be _possible_ to detect when multiprocessing is trying to spawn a resource tracker process and let it do its thing (i.e. very similar to what `freeze_support()` does):

```diff
diff --git a/nextstrain/cli/__main__.py b/nextstrain/cli/__main__.py
index 8af3c85..971d915 100644
--- a/nextstrain/cli/__main__.py
+++ b/nextstrain/cli/__main__.py
@@ -28,6 +28,12 @@ def main():
     if getattr(sys, "frozen", None):
         multiprocessing.spawn.freeze_support()
 
+        import re
+        for arg in sys.argv[1:]:
+            if re.search('^from multiprocessing[.]resource_tracker import main;main\([0-9]+\)$', arg):
+                exec(arg)
+                sys.exit()
+
     return cli.run( argv[1:] )
```

but that's getting increasingly gross.

**Comment by tsibley** [comment]
Instead of the complicated, intertwined patches above, I fixed this in #244 by relenting and using the default platform-dependent start methods. I added tests that exercise `open_browser()` under both "fork" and "spawn" to try to catch future bugs arising from the platform difference (in lieu of consistently using just "spawn").

**Comment by tsibley** [comment]
The fix will be part of 6.0.1, which is going thru CI now.

**Comment by tsibley** [comment]
After an issue with the release of 6.0.1, this is now released in 6.0.2.
